lib/rules/hook-use-state.js

```diff
@@ -0,0 +1,120 @@
+      category: 'Best Practices',
+      recommended: false,
+      url: docsUrl('hook-use-state'),
+    },
+    fixable: 'code',
+    messages: {
+      [USE_STATE_ERROR_MESSAGE]: 'useState call is not destructured into value + setter pair',
+    },
+    schema: [],
+  },
+
+  create: Components.detect((context) => {
+    let reactImportLocalName;
+    let reactUseStateLocalName;
+
+    return {
+      CallExpression(node) {
+        const isReactUseStateCall = (
+          reactImportLocalName
+          && node.callee.type === 'MemberExpression'
+          && node.callee.object.type === 'Identifier'
+          && node.callee.object.name === reactImportLocalName
+          && node.callee.property.type === 'Identifier'
+          && node.callee.property.name === 'useState'
+        );
+
+        const isUseStateCall = (
+          reactUseStateLocalName
+          && node.callee.type === 'Identifier'
+          && node.callee.name === reactUseStateLocalName
+        );
+
+        // Ignore unless this is a useState() or React.useState() call.
+        if (!isReactUseStateCall && !isUseStateCall) {
+          return;
+        }
+
+        const isDestructuringDeclarator = (
+          node.parent.type === 'VariableDeclarator'
+          && node.parent.id.type === 'ArrayPattern'
+        );
+
+        if (!isDestructuringDeclarator) {
+          context.report({ node, messageId: USE_STATE_ERROR_MESSAGE });
+          return;
+        }
+
+        const variableNodes = node.parent.id.elements;
+        const valueVariable = variableNodes[0];
+        const setterVariable = variableNodes[1];
+
+        const valueVariableName = valueVariable
+          ? valueVariable.name
+          : undefined;
+
+        const setterVariableName = setterVariable
+          ? setterVariable.name
+          : undefined;
+
+        const expectedSetterVariableName = valueVariableName ? (
+          `set${valueVariableName.charAt(0).toUpperCase()}${valueVariableName.slice(1)}`
+        ) : undefined;
+
+        if (
+          !valueVariable
+          || !setterVariable
+          || setterVariableName !== expectedSetterVariableName
+          || variableNodes.length !== 2
+        ) {
+          context.report({
+            node: node.parent.id,
+            messageId: USE_STATE_ERROR_MESSAGE,
+            fix: valueVariableName ? (fixer) => fixer.replaceTextRange(
+              [node.parent.id.range[0], node.parent.id.range[1]],
+              `[${valueVariableName}, ${expectedSetterVariableName}]`
+            ) : undefined,
+          });
+        }
+      },
+      ImportDeclaration(node) {
+        const isReactImported = node.source.type === 'Literal' && node.source.value === 'react';
+        const reactDefaultSpecifier = node.specifiers.find((specifier) => specifier.type === 'ImportDefaultSpecifier');
+        reactImportLocalName = reactDefaultSpecifier ? reactDefaultSpecifier.local.name : undefined;
```

Comment:
i think this stuff needs to be stored in eslint state, rather than in local variables, since rules can be run with an eslint server in an IDE, or in watch mode via jest-eslint-runner, etc.

Reply:
I extended the `Components.detect` work to track the names of React imports the same way it currently tracks the list of possible components.
It tracks the default React import as well as all named React imports; I only need the `useState` import for the feature in this PR, but a general approach seemed more appropriate.
The `hook-set-state-names` rule accesses the named imports and searches for the `useState` import.

I isolated the `Components.detect` changes to their own commit and rebased it to the head of this PR.
Let me know if you think this is the right approach, and if so, whether the `Components.detect` change should be its own PR, or should remain on this one.

Reply:
"own PR" was the right way :-)